File: DDD DSL/DDD DSL Design War FizzBuzzWhizz Reloaded/Code/Java/src/fayelab/ddd/fbwreloaded/interpreter/Interpreter.java
package fayelab.ddd.fbwreloaded.interpreter;

import java.util.Optional;

public class Interpreter
{
    public static Optional<String> interpret(Rule rule, int n)
    {
        return applyRule(rule, n);
    }

    private static Optional<String> applyRule(Rule rule, int n)
    {
        switch(rule.getType())
        {
            case ATOM:
                return applyAtom((Atom)rule.getData(), n);
            case OR:
                return applyOr((Or)rule.getData(), n);    
            case AND:
                return applyAnd((And)rule.getData(), n);
            default:
                return Optional.empty();
        }
    }

    private static Optional<String> applyAtom(Atom atom, int n)
    {
        if(applyPredication(atom.getPredication(), n))
        {
            return Optional.of(applyAction(atom.getAction(), n));
        }
        
        return Optional.empty();
    }

    private static boolean applyPredication(Predication predication, int n)
    {
        switch(predication.getType())
        {
            case TIMES:
                return times(predication.getParam(), n);
            case CONTAINS:
                return contains(predication.getParam(), n);
            case ALWAYSTRUE:
                return alwaysTrue();
            default:
                return false;
        }
    }

    private static String applyAction(Action action, int n)
    {
        switch(action.getType())
        {
            case TOFIZZ:
                return "Fizz";
            case TOBUZZ:
                return "Buzz";
            case TOWHIZZ:
                return "Whizz";
            case TOSTR:
                return String.valueOf(n);
            default:
                return "";
        }
    }
    
    private static Optional<String> applyOr(Or or, int n)
    {
        Optional<String> result1 = applyRule(or.getRule1(), n);
        if(result1.isPresent())
        {
            return result1;
        }
        
        return applyRule(or.getRule2(), n);
    }

    private static Optional<String> applyAnd(And and, int n)
    {
        Optional<String> result1 = applyRule(and.getRule1(), n);
        if(!result1.isPresent())
        {
            return Optional.empty();
        }
        
        Optional<String> result2 = applyRule(and.getRule2(), n);
        if(!result2.isPresent())
        {
            return Optional.empty();
        }
        
        return Optional.of(result1.get() + result2.get());
    }
    
    private static boolean times(int base, int n)
    {
        return n % base == 0;
    }

    private static boolean contains(int digit, int n)
    {
        int p1 = n % 10;
        int p2 = (n / 10) % 10;
        int p3 = (n / 10 / 10) % 10;
        return p1 == digit || p2 == digit || p3 == digit;
    }

    private static boolean alwaysTrue()
    {
        return true;
    }
}

File: DDD DSL/DDD DSL Design War FizzBuzzWhizz Reloaded/Code/Python/complete/interpreter/interpreter.py
def i_times(base, n):
    return n % base == 0

def i_contains(digit, n):
    return str(digit) in str(n)

def i_alwaystrue(param, n):
    return True

def i_tofizz(n):
    return 'Fizz'

def i_tobuzz(n):
    return 'Buzz'

def i_towhizz(n):
    return 'Whizz'

def i_tostr(n):
    return str(n)    

PREDICATION_MAP = {'TIMES':i_times, 'CONTAINS':i_contains, 'ALWAYSTRUE':i_alwaystrue}
ACTION_MAP = {'TOFIZZ':i_tofizz, 'TOBUZZ':i_tobuzz, 'TOWHIZZ':i_towhizz, 'TOSTR':i_tostr}

def apply_atom(predication, action, n):
    if apply_predication(predication, n):
        return True, apply_action(action, n)
    return False, ''

def apply_predication(predication, n):
    if predication[0] in PREDICATION_MAP:
        return PREDICATION_MAP[predication[0]](predication[1], n)
    return False

def apply_action(action, n):
    if action in ACTION_MAP:
        return ACTION_MAP[action](n)
    return ''

def apply_or(rule1, rule2, n):
    result1 = apply_rule(rule1, n)
    if result1[0]:
        return result1
    return apply_rule(rule2, n)

def apply_and(rule1, rule2, n):
    result1 = apply_rule(rule1, n)
    if not result1[0]:
        return False, ''
    result2 = apply_rule(rule2, n)
    if not result2[0]:
        return False, ''
    return True, ''.join([result1[1], result2[1]])

RULE_MAP = {'ATOM':apply_atom, 'OR':apply_or, 'AND':apply_and}

def apply_rule(rule, n):
    if rule[0] in RULE_MAP:
        return RULE_MAP[rule[0]](rule[1], rule[2], n)
    else:
        return False, ''
